drafts/str4d-overwinter-sighash/draft1.rst

```diff
@@ -0,0 +1,318 @@
+   input-output pairs, as long as each pair is located at an equivalent index.
+
+Field definitions
+-----------------
+
+The items 7, 9, 10a, 10d have the same meaning as the original algorithm from Bitcoin. [#wiki-checksig]_
+
+1: ``header``
+`````````````
+Deserialized into two transaction properties: ``fOverwintered`` and ``nVersion``. For transactions that use
+this transaction digest algorithm, ``fOverwintered`` is always set. [#ZIP-overwinter-tx-format]_
+
+2: ``nVersionGroupId``
+``````````````````````
+Provides domain separation of ``nVersion``. It is only defined if ``fOverwintered`` is set, which means that
+it is always defined for transactions that use this algorithm. [#ZIP-overwinter-tx-format]_
+
+3: ``hashPrevouts``
+```````````````````
+* If the ``ANYONECANPAY`` flag is not set, ``hashPrevouts`` is the BLAKE2b-256 hash of the serialization of
+  all input outpoints;
+
+* Otherwise, ``hashPrevouts`` is a ``uint256`` of ``0x0000......0000``.
+
+4: ``hashSequence``
+```````````````````
+* If none of the ``ANYONECANPAY``, ``SINGLE``, ``NONE`` sighash type is set, ``hashSequence`` is the
+  BLAKE2b-256 hash of the serialization of ``nSequence`` of all inputs;
+
+* Otherwise, ``hashSequence`` is a ``uint256`` of ``0x0000......0000``.
+
+5: ``hashOutputs``
+``````````````````
+* If the sighash type is neither ``SINGLE`` nor ``NONE``, ``hashOutputs`` is the BLAKE2b-256 hash of the
+  serialization of all output amount (8-byte little endian) with ``scriptPubKey`` (serialized as scripts
+  inside CTxOuts);
+
+* If sighash type is ``SINGLE`` and the input index is smaller than the number of outputs, ``hashOutputs`` is
+  the BLAKE2b-256 hash of the output (serialized as above) with the same index as the input;
+
+* Otherwise, ``hashOutputs`` is a ``uint256`` of ``0x0000......0000``. [#01-change]_
+
+6: ``hashJoinSplits``
+`````````````````````
+* If ``vjoinsplits`` is non-empty, ``hashJoinSplits`` is the BLAKE2b-256 hash of the serialization of all
+  JoinSplits (in their canonical transaction serialization format) concatenated with the joinSplitPubKey;
+
+  * Note that while signatures are ommitted, the JoinSplit proofs are included in the signature hash, as with
+    v1 and v2 transactions.
+
+* Otherwise, ``hashJoinSplits`` is a ``uint256`` of ``0x0000......0000``.
+
+8: ``nExpiryHeight``
+````````````````````
+The block height after which the transaction becomes unilaterally invalid, and can never be mined.
+[#ZIP-tx-expiry]_
+
+10b: ``scriptCode``
+```````````````````
+The script being currently executed: ``redeemScript`` for P2SH, or ``scriptPubKey`` in the general case. This
+is the same script as serialized in the Sprout transaction digest algorithm.
+
+10c: value
+`````````
+An 8-byte value of the amount of ZEC spent in this input.
+
+Notes
+-----
+
+When generating ``hashPrevouts``, ``hashSequence``, ``hashOutputs``, and ``hashJoinSplits``, the BLAKE2b-256
+personalization field is set to ``Zcash_Inner_Hash``.
```

Comment:
Is there a specific reason for sharing the personalization field across all 4 input types? It seems better to domain separate these different input types if possible. (To be clear I don't see any specific problem caused by this, and perhaps using a single string allows some optimization I've missed - but if so, maybe the text should mention said optimization.)

Reply:
Yes, let's change these to:
* ``ZcashPrevoutHash``
* ``ZcashSequencHash`` [spelling intended]
* ``ZcashOutputsHash``
* ``ZcashJSplitsHash``